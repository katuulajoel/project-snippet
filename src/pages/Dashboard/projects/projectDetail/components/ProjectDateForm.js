/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import PropTypes from "prop-types";
import React, { useState } from "react";
import { FormGroup } from "reactstrap";
import moment from "moment";
import styled from "styled-components";

/* -------------------------- Internel Dependencies ------------------------- */
import TextArea from "../../../../../components/TextArea";
import FieldError from "../../../../../components/FieldError";

/* --------------------------- Style dependencies --------------------------- */
import { StyledForm } from "../../../../../utils/styles";
import DateTimePicker from "../../../../../components/DateTimePicker";

const ProjectDateForm = (props) => {
  const { proceed, id } = props;

  const [project, setProject] = useState(props.project || {});
  const [errors, setErrors] = useState({ start_error: null, end_error: null });

  const onChangeValue = (key, value) => {
    let newState = {};
    newState[key] = value;
    setProject({ ...project, ...newState });
  };

  const onChangeField = (key, e) => {
    onChangeValue(key, e.target.value);
  };

  const onSave = (e) => {
    e.preventDefault();

    if (project.start_date == null) {
      setErrors({ start_error: "Start date is required" });
    } else {
      setErrors({ start_error: null });
    }
    if (project.deadline == null) {
      setErrors({ end_error: "End date is required" });
    } else {
      setErrors({ end_error: null });
    }

    if (proceed && project.start_date != null && project.deadline != null) {
      if (new Date(project.start_date) > new Date(project.deadline)) {
        setErrors({ start_error: "Start date is greater than end date" });
      } else {
        proceed(project);
      }
    }
  };

  return (
    <StyledForm id={id} onSubmit={onSave}>
      <label>
        Timeline
        <LabelStyle>
          <span className="label-style">*</span>
        </LabelStyle>
      </label>
      <div style={{ display: "flex", gap: "14px" }}>
        <FormGroup>
          {errors.start_error && <FieldError message={errors.start_error} />}
          <DateTimePicker
            className="tg-date-field"
            placeholder="Enter Start Date"
            format={"DD MMM YYYY"}
            calendar={true}
            time={false}
            value={project.start_date ? new Date(project.start_date) : null}
            onChange={(date) => {
              onChangeValue("start_date", moment.utc(date).format());
            }}
            required
          />
        </FormGroup>
        <FormGroup>
          {errors.end_error && <FieldError message={errors.end_error} />}
          <DateTimePicker
            className="tg-date-field"
            placeholder="Enter Deadline"
            format={"DD MMM YYYY"}
            calendar={true}
            time={false}
            value={project.deadline ? new Date(project.deadline) : null}
            onChange={(date) => {
              onChangeValue("deadline", moment.utc(date).format());
            }}
            required
          />
        </FormGroup>
      </div>

      {project.start_date && project.deadline ? (
        <FormGroup>
          <label>
            Give reason for change
            <LabelStyle>
              <span className="label-style">*</span>
            </LabelStyle>
          </label>
          <TextArea
            placeholder="Enter reason here"
            value={project.reason}
            onChange={(e) => onChangeField("reason", e)}
            required
          />
        </FormGroup>
      ) : null}
    </StyledForm>
  );
};

const LabelStyle = styled.span`
  .label-style {
    color: #da3451;
    padding-left: 2px;
  }
`;

ProjectDateForm.propTypes = {
  project: PropTypes.shape({
    id: PropTypes.number,
    start_date: PropTypes.string,
    deadline: PropTypes.string,
    reason: PropTypes.string,
  }),
  proceed: PropTypes.func,
  cancel: PropTypes.func,
  dismiss: PropTypes.func,
  id: PropTypes.string,
};

export default ProjectDateForm;
