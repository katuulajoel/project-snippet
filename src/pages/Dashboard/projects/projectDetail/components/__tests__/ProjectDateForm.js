/* eslint-disable react/prop-types */
import React from "react";
import renderer from "react-test-renderer";
import ProjectDateForm from "../ProjectDateForm";
import { mount } from "enzyme/build";

function mockComponent(componentName) {
  return (props) => {
    return (
      <input
        originalComponent={componentName}
        {...props}
        onChange={(e) => {
          props.onChange(e.target.value);
        }}
      />
    );
  };
}

jest.mock("../../../../../../components/DateTimePicker", () => {
  return mockComponent("DateTimePicker");
});

describe("Project Planning test", () => {
  beforeAll(() => {
    Date.now = jest.fn(() => new Date(Date.UTC(2017, 1, 14)).valueOf());
  });
  it("Should match snapshot test", () => {
    const tree = renderer.create(<ProjectDateForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should save form if all required fields are correct", () => {
    const proceedStub = jest.fn();
    const wrapper = mount(
      <ProjectDateForm project={{ id: 123 }} proceed={proceedStub} />
    );

    wrapper.find("form").simulate("submit");

    expect(wrapper.find("FieldError").props().message).toBe(
      "End date is required"
    );

    wrapper
      .find("input")
      .at(0)
      .simulate("change", { target: { value: "2020-03-29T10:12:33.220000" } });
    wrapper
      .find("input")
      .at(1)
      .simulate("change", { target: { value: "2020-03-28T10:12:33.220000" } });

    wrapper.find("form").simulate("submit");

    expect(wrapper.find("FieldError").props().message).toBe(
      "Start date is greater than end date"
    );

    wrapper
      .find("input")
      .at(0)
      .simulate("change", { target: { value: "2020-03-28T10:12:33.220000" } });
    wrapper
      .find("input")
      .at(1)
      .simulate("change", { target: { value: "2020-03-29T10:12:33.220000" } });
    wrapper
      .find("textarea")
      .simulate("change", { target: { value: "description" } });

    wrapper.find("form").simulate("submit");

    expect(proceedStub).toHaveBeenCalledWith({
      start_date: "2020-03-28T10:12:33Z",
      deadline: "2020-03-29T10:12:33Z",
      id: 123,
      reason: "description",
    });
  });
});
