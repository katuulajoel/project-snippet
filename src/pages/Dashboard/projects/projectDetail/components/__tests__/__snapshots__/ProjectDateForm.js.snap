// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Planning test Should match snapshot test 1`] = `
<form
  className="sc-gtsrHT jKtEre"
  onSubmit={[Function]}
>
  <label>
    Timeline
    <span
      className="sc-eCApnc gXWHPB"
    >
      <span
        className="label-style"
      >
        *
      </span>
    </span>
  </label>
  <div
    style={
      Object {
        "display": "flex",
        "gap": "14px",
      }
    }
  >
    <div
      className="form-group"
    >
      <input
        calendar={true}
        className="tg-date-field"
        format="DD MMM YYYY"
        onChange={[Function]}
        originalComponent="DateTimePicker"
        placeholder="Enter Start Date"
        required={true}
        time={false}
        value={null}
      />
    </div>
    <div
      className="form-group"
    >
      <input
        calendar={true}
        className="tg-date-field"
        format="DD MMM YYYY"
        onChange={[Function]}
        originalComponent="DateTimePicker"
        placeholder="Enter Deadline"
        required={true}
        time={false}
        value={null}
      />
    </div>
  </div>
</form>
`;
